crates/ruff_linter/src/rules/flake8_quotes/rules/avoidable_escaped_quote.rs

```diff
@@ -0,0 +1,258 @@
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::str::{is_triple_quote, leading_quote};
+use ruff_python_parser::lexer::LexResult;
+use ruff_python_parser::Tok;
+use ruff_source_file::Locator;
+use ruff_text_size::TextRange;
+
+use crate::lex::docstring_detection::StateMachine;
+use crate::registry::AsRule;
+use crate::settings::LinterSettings;
+
+/// ## What it does
+/// Checks for strings that include escaped quotes, and suggests changing
+/// the quote style to avoid the need to escape them.
+///
+/// ## Why is this bad?
+/// It's preferable to avoid escaped quotes in strings. By changing the
+/// outer quote style, you can avoid escaping inner quotes.
+///
+/// ## Example
+/// ```python
+/// foo = 'bar\'s'
+/// ```
+///
+/// Use instead:
+/// ```python
+/// foo = "bar's"
+/// ```
+#[violation]
+pub struct AvoidableEscapedQuote;
+
+impl AlwaysAutofixableViolation for AvoidableEscapedQuote {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Change outer quotes to avoid escaping inner quotes")
+    }
+
+    fn autofix_title(&self) -> String {
+        "Change outer quotes to avoid escaping inner quotes".to_string()
+    }
+}
+
+struct FStringContext {
+    /// Whether to check for escaped quotes in the f-string.
+    check_for_escaped_quote: bool,
+    /// The range of the f-string start token.
+    fstring_start_range: TextRange,
+    /// The ranges of the f-string middle tokens containing escaped quotes.
+    fstring_middle_ranges: Vec<TextRange>,
+}
+
+impl FStringContext {
+    fn new(check_for_escaped_quote: bool, fstring_start_range: TextRange) -> Self {
+        Self {
+            check_for_escaped_quote,
+            fstring_start_range,
+            fstring_middle_ranges: vec![],
+        }
+    }
+
+    /// Update the context to not check for escaped quotes, and clear any
+    /// existing reported ranges.
+    fn do_not_check_for_escaped_quote(&mut self) {
+        self.check_for_escaped_quote = false;
+        self.fstring_middle_ranges.clear();
+    }
```

Comment:
I would have expected a `enter_`...`exit_...` function pair How does it start checking for quotes again?

Reply:
We don't start again. For example,

```python
f"\"foo\" {'nested'}"
# ^^(1)^^^ ^^^(2)^^
```

Here, (1) will be checked first and the range will be added to `fstring_middle_ranges`. Then we encountered a nested string (2) but our target version doesn't support PEP 701, so we'll stop checking for the current f-string by inverting the flag and removing any previously diagnosed ranges.